Modules/Filtering/Convolution/include/itkFFTConvolutionImageFilter.hxx

```diff
@@ -126,47 +147,154 @@ FFTConvolutionImageFilter<TInputImage, TKernelImage, TOutputImage, TInternalPrec
   ProgressAccumulator *      progress,
   float                      progressWeight)
 {
-  // Pad the image
-  InputSizeType   padSize = this->GetPadSize();
-  InputRegionType inputRegion = input->GetLargestPossibleRegion();
-  InputSizeType   inputSize = inputRegion.GetSize();
+  float remainingProgress = 1.0f;
+
+  InputRegionType  inputLargestRegion = input->GetLargestPossibleRegion();
+  InputSizeType    inputLargestSize = inputLargestRegion.GetSize();
+  InputIndexType   inputLargestIndex = inputLargestRegion.GetIndex();
+  InputRegionType  inputRequestedRegion = input->GetRequestedRegion();
+  InputSizeType    inputRequestedSize = inputRequestedRegion.GetSize();
+  InputIndexType   inputRequestedIndex = inputRequestedRegion.GetIndex();
+  OutputRegionType outputRequestedRegion = this->GetOutput()->GetRequestedRegion();
+  OutputSizeType   outputRequestedSize = outputRequestedRegion.GetSize();
+  OutputIndexType  outputRequestedIndex = outputRequestedRegion.GetIndex();
+
+  // Pad the input image such that the requested region, expanded by
+  // twice the kernel radius, lies entirely within the buffered region.
+  // If the requested region is at least the kernel radius away from
+  // all boundaries of the buffered input region then padding is skipped.
 
   using InputPadFilterType = PadImageFilter<InputImageType, InputImageType>;
-  auto inputPadder = InputPadFilterType::New();
-  inputPadder->SetBoundaryCondition(this->GetBoundaryCondition());
+  using PadSizeType = typename InputPadFilterType::SizeType;
 
-  InputSizeType inputLowerBound = this->GetPadLowerBound();
-  inputPadder->SetPadLowerBound(inputLowerBound);
+  PadSizeType    lowerPad, upperPad;
+  bool           needsKernelPadding = false;
+  KernelSizeType kernelRadius = this->GetKernelRadius();
+  for (unsigned int dim = 0; dim < ImageDimension; ++dim)
```

Comment:
What is the intent here for even and odd kernel sizes?  Is it that if the kernel size is `2r+1` then we are growing a range `[i,i+s)` to `[i-r,i+s+r)`, which means we need to pad if `i-r < inputLargestIndex` or if `i+s+r > inputLargestIndex+inputLargestSize`?

What happens if the kernel size is `2r`?  Do we grow the region the same, as if the kernel size were actually `2r+1`?  Or, for example, do we instead grow the upper limit by `r-1`?

Reply:
We are padding by `r` on each side, meaning that the resulting region has size `size[dim] + 2 * r[dim]` in each direction regardless of whether the kernel is of size `2r` or `2r+1`. So, yes, we treat even-sized kernel images as if the kernel size was actually `2r+1` as you've said.